frontends/election-manager/src/components/format_usb_modal.tsx

```diff
@@ -0,0 +1,178 @@
+import React, { useCallback, useContext, useEffect, useState } from 'react';
+import { assert, throwIllegalValue } from '@votingworks/utils';
+
+import {
+  Button,
+  Modal,
+  Prose,
+  isElectionManagerAuth,
+  isSystemAdministratorAuth,
+  LinkButton,
+  Loading,
+  UsbImage,
+} from '@votingworks/ui';
+import { AppContext } from '../contexts/app_context';
+
+export interface FormatUsbModalProps {
+  onClose: () => void;
+}
+
+type ModalState = 'init' | 'confirm' | 'formatting' | 'done' | 'error';
+
+export function FormatUsbModal({ onClose }: FormatUsbModalProps): JSX.Element {
+  const { usbDrive, auth } = useContext(AppContext);
+  assert(isSystemAdministratorAuth(auth) || isElectionManagerAuth(auth));
+  const userRole = auth.user.role;
+
+  const [state, setState] = useState<ModalState>('init');
+  const [errorMessage, setErrorMessage] = useState('');
```

Comment:
One best practice I've learned is to model state that changes together as one state object. In this case, there is an implicit invariant that `errorMessage` is only non-empty when `state` is not `'error'`. We can build that invariant into the data structure to help enforce it:

```ts
type ModalState =
  | { state: 'init' }
  | { state: 'confirm' }
  ...
  | { state: 'error', message: string }
```

In this particular case, it's not so complicated so it's not the most important, but thought I'd mention it because I think it's a good habit in general.

Reply:
Changed.